### PR TITLE
Add border/divider options to InfoUnitGroup

### PR DESCRIPTION
**Pull request by virginiacc**
Adds border/divider options to InfoUnitGroup

## Additions

- `has_top_border` and `has_bottom_border` options on InfoUnitGroup

## Changes

- Extends `lines_between_items` to all InfoUnitGroup types using `content-l__top-divider` class. Dividers display between `content-l`s on desktop and also between `content-l-col`s on mobile.

## Testing

1. Check out branch and run migrations & `gulp styles`
2. In admin, create a page with InfoUnitGroups of each type. Select `lines_between_items`, `has_top_border`, and/or `has_bottom_border`  on each InfoUnitGroup. Preview page and verify that dividers display as shown in screenshots.

## Notes
- `content-l__top-divider` might be a little deceptive since the dividers appear between units. I'm thinking about changing the name -- or maybe moving the modifier up to the info unit group.

### Borders/dividers on desktop
screencapture-localhost-8000-admin-pages-10686-edit-preview-1507736148433

### Dividers on mobile 

screen shot 2017-10-12 at 7 47 33 am

## Checklist

* [ ] PR has an informative and human-readable title
* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [development playbook](https://github.com/cfpb/development)
* [ ] Passes all existing automated tests
* [ ] Any *change* in functionality is tested
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated
* [ ] Reviewers requested with the [Reviewer tool](https://help.github.com/articles/about-pull-request-reviews/) :arrow_right:

### CODE REVIEW

File: cfgov/unprocessed/css/organisms/info-unit-group.less [hunk 1]
```diff
@@ -63,8 +63,33 @@
...
+                margin-top: unit( @grid_gutter-width / @base-font-size-px, em );
+            }
+        }
+    });
 
     .respond-to-min( @bp-sm-min, {
+        .content-l__top-divider + .content-l__top-divider {
+            // Matches top divider width to cols on larger screens
+            &:before {
+                margin-left: 15px !important;
```
Comment (anselmbradford) [inline cfgov/unprocessed/css/organisms/info-unit-group.less]: Add note as to why `!important` is necessary?

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html [hunk 2]
```diff
@@ -58,20 +64,36 @@
...
     {{ value.intro | safe }}
 {%- endif %}
-
-    <div class="content-l">
+  {% if not value.lines_between_items %}
+  <div class="content-l">
+  {% endif %}
     {% for unit in value.info_units %}
+        {% if value.lines_between_items and  (
+          ( loop.index == 1 ) or
```
Comment (jimmynotjim) [inline cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Minor, Indentation should be four spaces.

Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Should be fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/cd56484fa764a5930cc8390511c7c89ca8bb1889.

File: cfgov/unprocessed/css/organisms/info-unit-group.less (hunk 1 above)
Comment (virginiacc) [reply cfgov/unprocessed/css/organisms/info-unit-group.less]: Note added in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/cd56484fa764a5930cc8390511c7c89ca8bb1889.

**Comment by Scotchester** [comment]
Please hold this until I can review. I want to double-check something.

**Comment by Scotchester** [comment]
OK, the implementation seems fine to me.

The concern I have is with enforcing our standards for how rule lines should be used to divide sections. See: https://[GHE]/flapjack/Modules-V1/wiki/Atoms#grey-rule-line

Specifically, when using one line between major page sections, the line should be 60px from the bottom of the top section and 30px from the top of the bottom section. If we allow both top & bottom rule lines, using all bottom rule lines instead of tops would put the line only 30px from the bottom of the top section, and 60px from the top of the bottom section, as in Virginia's screenshot.

Sorry, I know this is hard to follow. To visually explain it, this is how (as I understand it) a page laid out like Virgina's example should look:

screen shot 2017-10-16 at 11 33 49-fullpage

Can we remove the option to add an overall bottom border and only have an option for top borders? Or is there another use case for overall bottom borders I'm missing?

/cc @stephanieosan

**Comment by virginiacc** [comment]
@Scotchester Ah, good point, thanks! I discussed with the team and we think the top border option should be sufficient, so I'll remove the other one.

**Comment by Scotchester** [comment]
Thanks, @virginiacc.

The other thing I'd like to address (or start to address) here is our inconsistent use of "rule", "line", and "border".

- Text Introductions have a "Has rule" field.
- The existing Link Blob modules used "Has top/bottom border" fields
- The new Info Unit Group module has a "Show lines between items" field
- These are typically referred to in conversation as "rule lines", a phrase which appears in none of the above, but a phrase that might be considered jargon and might be unfamiliar to a content editor, or to a developer working on modifying these features in the future.

I don't think we need to completely solve this across all modules in this PR, but we should probably at least standardize the two fields in this module before merging this.

Finally, we should probably also hold this for a response from JP on related issue https://[GHE]/CFPB/hubcap/issues/55

**Comment by virginiacc** [comment]
@Scotchester I agree that it makes sense to standardize the terminology for the lines. I'd favor using `rule lines` since that's what we generally seem to call them, or maybe `divider lines` if that sounds more accessible?

**Comment by Scotchester** [comment]
Sounds good to me, @virginiacc. It shouldn't take much education to get our regular content editors familiar with the term, if they aren't already. Per JP's response  in hubcap#55, would you like to add the option to our other major section modules in this PR, as well?

**Comment by virginiacc** [comment]
@Scotchester Ok, thanks! Yes, I'll push another commit to add the rule options to the other modules.

File: cfgov/v1/atomic_elements/organisms.py [hunk 3]
```diff
@@ -913,6 +983,13 @@ class Media:
 class SnippetList(blocks.StructBlock):
     heading = blocks.CharBlock(required=False)
     body = blocks.RichTextBlock(required=False)
+    has_top_border = blocks.BooleanBlock(
+        default=False,
+        required=False,
+        label="Has top rule line",
+        help_text=('Check this to add a horizontal rule line to top of '
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: Should say `snippet list` instead of `chart block` here.

**Comment by virginiacc** [comment]
@Scotchester I think all the changes we discussed should be implemented now!

File: cfgov/v1/atomic_elements/organisms.py [hunk 4]
```diff
@@ -696,9 +710,14 @@ class Expandable(BaseExpandable):
...
-
     is_accordion = blocks.BooleanBlock(required=False)
-    has_rule = blocks.BooleanBlock(required=False)
+    has_rule = blocks.BooleanBlock(
+        default=False,
+        required=False,
+        label="Has top rule line",
+        help_text=('Check this to add a horizontal rule line to top of '
+                   'expandable group.')
+    )
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: As it turns out, due to a series of unfortunate events, this field actually triggers a rule line both above _and_ below the ExpandableGroup module. Without getting into the aforementioned data migration issue, the fix here is straightforward, but hard to explain in a comment here. Let's pair on it tomorrow or Thursday.

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html [hunk 5]
```diff
@@ -11,6 +11,9 @@
    value:                         Object defined from a StreamField block.
 
    value.format:                  The number and width of columns to output.
+   
+   value.has_top_border:          Boolean for whether or not to add top
+                                  border to info-unit-group
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: It might be worth noting here (and in the other templates for the organisms we're adding the option to) that this field is used by `render_block.html` to modify the wrapping `div.block`.

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html [hunk 6]
```diff
@@ -58,20 +61,36 @@
...
+            ( value.format == '33-33-33' and loop.index % 3 == 0 ) or
+            ( loop.index == value.info_units | length )
+        ) %}
+        </div>
+        {% endif %}
     {% endfor %}
+    {% if not value.lines_between_items %}
     </div>
-
+    {% endif %}
```
[the remark this answers is not in the document]
Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: I agree about the logic. What if we had a single loop that output the `content-l`s in all cases based on the index/format checks, and conditionally added the `content-l__top-divider` class if `lines_between_items` == true?

File: cfgov/v1/atomic_elements/organisms.py (hunk 4 above)
Comment (virginiacc) [reply cfgov/v1/atomic_elements/organisms.py]: That is unfortunate. :) What are you thinking for the fix? I checked the live instances of ExpandableGroup on cf.gov (there aren't too many) and it looks like we only use the 'has_rule' option [on one page](https://www.consumerfinance.gov/about-us/events/request-speaker/). I'm wondering, could we replace `has_rule` with `has_top_rule_line` here and then update the affected page once the change is deployed?

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html (hunk 5 above)
Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Updated in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/082d1b1bd65d8496ff7e23272d60267f81c565d4

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: This should be fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/082d1b1bd65d8496ff7e23272d60267f81c565d4. Thanks for the suggestion -- the logic seems much cleaner now.

File: cfgov/v1/atomic_elements/organisms.py [hunk 7]
```diff
@@ -64,12 +64,20 @@ class InfoUnitGroup(blocks.StructBlock):
                    'the first link in their unit\'s list, if there is a link.')
     )
 
+    has_top_border = blocks.BooleanBlock(
```
[the remark this answers is not in the document]
Comment (virginiacc) [reply cfgov/v1/atomic_elements/organisms.py]: Should be fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/082d1b1bd65d8496ff7e23272d60267f81c565d4

File: cfgov/v1/atomic_elements/organisms.py (hunk 4 above)
Comment (Scotchester) [reply cfgov/v1/atomic_elements/organisms.py]: Oh! I didn't even think to see how much it was being used. That's a much better idea than the one I had, which was to add more logic to `render_block.html`.

_(Also, I bet if the rule line on that page just went away, no one would notice 😁 …)_

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html [hunk 8]
```diff
@@ -59,19 +64,28 @@
...
-        {% if value.lines_between_items and value.format == '25-75' and loop.index > 1 %}
-            <div class="a-rule-break"></div>
-        {% endif %}
-            {{- info_unit(unit, value.format, value.link_image_and_heading) }}
-        </div>
-    {% endfor %}
+{% for unit in value.info_units %}
+    {% if ( loop.index == 1 ) or
+          ( value.format == '50-50' and loop.index % 2 == 1 ) or
+          ( value.format == '33-33-33' and loop.index  % 3 == 1 )
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Trivial: extra space before the modulo operator here.

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html [hunk 9]
```diff
@@ -59,19 +64,28 @@
...
+    <div class="content-l
+                {{' content-l__top-divider' if value.lines_between_items else ''}}">
+    {% endif %}
+    
+    <div class="content-l_col
+                {{- ' content-l_col-1-2' if value.format == '50-50' else '' -}}
+                {{- ' content-l_col-1-3' if value.format == '33-33-33' else '' -}}
+                {{- ' content-l_col-1' if value.format == '25-75' else '' -}}">
+        {{- info_unit(unit, value.format, value.link_image_and_heading) }}
     </div>
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Trivial: This `div` should be indented one more level

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html (hunk 8 above)
Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Should be fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/9f21d322fc70c2ba1ebd623972e54df7c0087e47

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html (hunk 9 above)
Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Should be fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/9f21d322fc70c2ba1ebd623972e54df7c0087e47

File: cfgov/v1/atomic_elements/organisms.py (hunk 4 above)
Comment (virginiacc) [reply cfgov/v1/atomic_elements/organisms.py]: @Scotchester That is probably true! 😄 

I updated `has_rule` to `has_top_rule_line` on ExpandableGroup in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/a67606e1c8d308c1f872928868ef0b08a2e08e6b. Thanks for noticing this issue!

File: cfgov/unprocessed/css/organisms/info-unit-group.less [hunk 10]
```diff
@@ -58,13 +58,50 @@
...
+    
+    .respond-to-max( @bp-xs-max, {
+        // Adds top divider between info unit cols on small screens
+        .content-l__top-divider {
+            .content-l_col + .content-l_col {
+                .grid_column__top-divider;
+                margin-top: unit( @grid_gutter-width / @base-font-size-px, em );
+            }
+        }
+    });
```
Comment (jimmynotjim) [inline cfgov/unprocessed/css/organisms/info-unit-group.less]: Minor, add space within paren here.

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html [hunk 11]
```diff
@@ -59,19 +64,28 @@
...
 
-    <div class="content-l">
-    {% for unit in value.info_units %}
+{% for unit in value.info_units %}
+    {% if ( loop.index == 1 ) or
+          ( value.format == '50-50' and loop.index % 2 == 1 ) or
+          ( value.format == '33-33-33' and loop.index % 3 == 1 ) 
+    %}
+    <div class="content-l
+                {{' content-l__top-divider' if value.lines_between_items else ''}}">
```
Comment (jimmynotjim) [inline cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Suggestion, using `{{-` here would eliminate the extra whitespace in the rendered html.

File: cfgov/v1/atomic_elements/organisms.py [hunk 12]
```diff
@@ -319,7 +319,7 @@ class Meta:
 
 
 class MainContactInfo(blocks.StructBlock):
-    contact = SnippetChooserBlock(ContactSnippetClass)
+    contact = SnippetChooserBlock('v1.Contact')
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: What's this change about?

**Comment by Scotchester** [comment]
Code changes look good, but you'll need to re-create your migration again, unfortunately. Should be 0099 now.

File: cfgov/unprocessed/css/organisms/info-unit-group.less (hunk 10 above)
Comment (virginiacc) [reply cfgov/unprocessed/css/organisms/info-unit-group.less]: Should be fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/4c649330961585504690e2ccfe018ae41b4e57cc

File: cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html (hunk 11 above)
Comment (virginiacc) [reply cfgov/jinja2/v1/_includes/organisms/info-unit-group-2.html]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/4c649330961585504690e2ccfe018ae41b4e57cc -- thanks!

**Comment by virginiacc** [comment]
@Scotchester Migrations updated in https://github.com/cfpb/cfgov-refresh/pull/3476/commits/023b5ba9c25bc0b398fcb8448b75f0d18ae45645.

**Comment by Scotchester** [comment]
Everything's working great! :shipit:

**Comment by virginiacc** [comment]
Merged before anything else changes. :)

Thanks for all the reviews/suggestions, @Scotchester, @jimmynotjim, and @anselmbradford!
